File: app/src/main/java/com/example/weatherapplication/SearchFragment/SearchFragment.java
package com.example.weatherapplication.SearchFragment;

import android.os.Bundle;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.Navigation;

import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.AutoCompleteTextView;
import android.widget.TextView;
import com.airbnb.lottie.LottieAnimationView;
import com.example.weatherapplication.AllCityInTheWorld;
import com.example.weatherapplication.CityFragment.CityFragment;
import com.example.weatherapplication.MainActivityViewModel;
import com.example.weatherapplication.R;
import com.example.weatherapplication.WatchList.WatchListDouInterface;
import com.example.weatherapplication.WatchList.WatchListWeather;

import static android.view.View.GONE;

public class SearchFragment extends Fragment{
    private AppCompatImageView backButton;
    private AutoCompleteTextView searchTextView;
    private SearchTextViewAdapter adapter;
    private MainActivityViewModel mainActivityViewModel;
    private View layout;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,Bundle savedInstanceState) {
        layout = inflater.inflate(R.layout.fragment_search, container, false);
        backButton = layout.findViewById(R.id.search_fragment_back);
        return layout;
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        mainActivityViewModel = new ViewModelProvider(requireActivity()).get(MainActivityViewModel.class);

        setAutoCompleteSearch(view);

        chooseSuggestion();

        View view3 = view.findViewById(R.id.free_space_suggestion);
        view3.setOnClickListener(view1 -> searchTextView.showDropDown());

        backButtonManaging();
    }

    public void setAutoCompleteSearch(View view){
        AllCityInTheWorld suggestionCities = new AllCityInTheWorld(getContext());
        searchTextView = view.findViewById(R.id.search_fragment_city_name);
        adapter = new SearchTextViewAdapter(getContext(),suggestionCities.getAllCitiesInTheWorld());
        searchTextView.setAdapter(adapter);
    }

    public void chooseSuggestion(){
        searchTextView.setOnItemClickListener((parent, view, position, id) -> {
            String cityName = getCityNameFromAutoCompleteTextView(view);
            double lat = getLatFromInformation(view);
            double lon = getLonFromInformation(view);
            String country = getCountryName(view);

            setAutoCompleteSearchTextView(cityName);
            mainActivityViewModel.setCityInformation(lat,lon,cityName,country);
            mainActivityViewModel.getIsNavigateFromSearchFragment().setValue(true);
            Navigation.findNavController(requireActivity(),R.id.fragmentContainerView).navigate(R.id.action_searchFragment_to_cityFragment2);

        });
        searchTextView.setOnKeyListener((v, keyCode, event) -> false);

    }

    public String getCityNameFromAutoCompleteTextView(View view){
        TextView cityDataInfoTV = view.findViewById(R.id.search_suggestion_city_name);
        return cityDataInfoTV.getText().toString().split("!")[0];
    }

    public double getLatFromInformation(View view){
        TextView latTv = view.findViewById(R.id.search_suggestion_city_other_information);
        return Double.parseDouble(latTv.getText().toString().split("/")[1].split(",")[0]);
    }

    public double getLonFromInformation(View view){
        TextView lonTv = view.findViewById(R.id.search_suggestion_city_other_information);
        return Double.parseDouble(lonTv.getText().toString().split("/")[1].split(",")[1]);
    }

    public String getCountryName(View view){
        TextView country = view.findViewById(R.id.search_suggestion_city_other_information);
        return country.getText().toString().split("/")[0];
    }

    public void setAutoCompleteSearchTextView(String cityName){
        searchTextView.setText(cityName);
    }

    public void backButtonManaging(){
        backButton.setOnClickListener(v -> getActivity().getSupportFragmentManager().popBackStack());
    }
}

File: app/src/main/java/com/amirreza/presentation/weatherapplication/SearchFragment/SearchFragment.kt
package com.amirreza.presentation.weatherapplication.SearchFragment

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.appcompat.widget.AppCompatTextView
import androidx.fragment.app.Fragment
import androidx.navigation.Navigation.findNavController
import com.amirreza.common.AllCityInTheWorld
import com.amirreza.domain.entity.SavedCityWeather
import com.amirreza.presentation.weatherapplication.CityFragment.CityFragmentViewModel
import com.amirreza.weatherapplication.R
import com.amirreza.weatherapplication.databinding.FragmentSearchBinding
import org.koin.androidx.viewmodel.ext.android.sharedViewModel

class SearchFragment:Fragment() {
    private lateinit var binding: FragmentSearchBinding
    private val cityFragmentViewModel by sharedViewModel<CityFragmentViewModel>()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        binding = FragmentSearchBinding.inflate(inflater,container,false)
        return binding.root
    }


    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        setAutoCompleteSearch()
        binding.freeSpaceSuggestion.setOnClickListener{
            binding.searchFragmentCityName.showDropDown()
        }
        backButtonManaging()
    }


    private fun setAutoCompleteSearch() {
        val suggestionCities = AllCityInTheWorld(requireContext())
        val adapter = SearchTextViewAdapter(
            requireContext(),
            suggestionCities.allCitiesInTheWorld,
            object : ItemClickListener{
                override fun onClick(view: View) {
                    val cityName: String = view.findViewById<AppCompatTextView>(R.id.search_suggestion_city_name).text as String
                    val country: String = getCountryName(view)
                    val lat = getLatFromInformation(view)
                    val lon = getLonFromInformation(view)

                    setAutoCompleteSearchTextView(country)


                    cityFragmentViewModel
                        .addCityToDatabaseFromSearchFragment(
                            SavedCityWeather(
                                cityName = cityName,
                                country = country,
                                description = "",
                                temperature = "",
                                weatherImagePath = "",
                                lat = lat,
                                lon = lon
                            )
                        )



                    findNavController(requireView()).popBackStack()
                }
            }
        )
        binding.searchFragmentCityName.setAdapter(adapter)
    }

    private fun getLatFromInformation(view: View): Double {
        val latTv = view.findViewById<TextView>(R.id.search_suggestion_city_other_information)
        return latTv.text.toString().split("/").toTypedArray()[1].split(",")
            .toTypedArray()[0].toDouble()
    }


    private fun getLonFromInformation(view: View): Double {
        val lonTv = view.findViewById<TextView>(R.id.search_suggestion_city_other_information)
        return lonTv.text.toString().split("/").toTypedArray()[1].split(",")
            .toTypedArray()[1].toDouble()
    }

    private fun getCountryName(view: View): String {
        val country = view.findViewById<TextView>(R.id.search_suggestion_city_other_information)
        return country.text.toString().split("/").toTypedArray()[0]
    }
    private fun setAutoCompleteSearchTextView(cityName: String?) {
        binding.searchFragmentCityName.setText(cityName)
    }
    private fun backButtonManaging() {
        binding.searchFragmentBack.setOnClickListener {
            requireActivity().supportFragmentManager.popBackStack()
        }
    }
}
interface ItemClickListener{
    fun onClick(view: View)
}